### PR TITLE
Introduce the API feature to BitcoinD.

### LINKED ISSUE #1685: Add Api support to BitcoinD
At the moment, for some unknown reason, bitcoinD isn't using the Api feature.

### PR DESCRIPTION
**Pull request by ferdeen**
Fix #1685.

This was required so I can test syncing wallets ( #1103) when running BitcoinD. 

Confirm Api support :

image

And in Swagger :

image

### CODE REVIEW

**Comment by noescape00** [comment]
Does all the API methods support btc network? Can you please check if every method works appropriately and doesn't throw? 
I'd expect potential problems with methods like getstakinginfo/startstaking/stopstaking. Also those methods should not be present in the API in case it's running on BTC network.

File: src/Stratis.BitcoinD/Program.cs [hunk 1]
```diff
@@ -29,6 +30,7 @@ public class Program
                     .AddMining()
                     .AddRPC()
                     .UseWallet()
+                    .UseApi()
```
Comment (dangershony) [inline src/Stratis.BitcoinD/Program.cs]: Could we also add apps in that case?

**Comment by dangershony** [comment]
@noescape00 I am not sure you will have those methods as they are not registered in the BTC mining feature.

File: src/Stratis.BitcoinD/Program.cs (hunk 1 above)
Comment (ferdeen) [reply src/Stratis.BitcoinD/Program.cs]: Yes will do.

**Comment by ferdeen** [comment]
Tested the majority of the web api calls (@dangershony & @noescape00).

- GetRawMempool didn't return an empty value
- StartStaking failed 
- The rest were fine - I didn't fully test all the wallet api call, as the majority are working as normal.

@dangershony  For the failures do we remove the web api calls ?  I don't want to spend too much time on this as this card was included as a quick win and simple task.

Test results in the below image.

image

**Comment by noescape00** [comment]
Ok. So there are staking related api methods on BTC network? 
Those should be disabled.

**Comment by ferdeen** [comment]
I am unable to hide the MiningFeature web API calls by network.  However, I have returned early out of the calls when on the the Bitcoin network :

image

**Comment by ferdeen** [comment]
Sorry hold off, can see some related tests are failing.  Let me look into this.  Cheers.

**Comment by noescape00** [comment]
>I am unable to hide the MiningFeature web API calls by network. However, I have returned early out of the calls when on the the Bitcoin network :

Check this out: 
https://stackoverflow.com/questions/38991603/how-can-i-disable-some-apis-of-my-asp-net-application

We can introduce attributes like [StratisOnly] and use them on staking methods.

**Comment by ferdeen** [comment]
Trying out the `ActionFilterAttribute`.  

Problem I'm facing is trying to pass `this.fullNode...` as a parameter to the attribute.  I need to know what the network is to determine whether to bypass the action.

image

For now can we leave it with the check within the method calls, as this is stopping me from completing a more important task - #1103 ?

**Comment by noescape00** [comment]
> as this is stopping me from completing a more important task - #1103 ?

Why? 1103 has nothing to do with API

>For now can we leave it with the check within the method calls

I don't mind. Create an issue for that and assign it to yourself and let's for now have current approach.

**Comment by ferdeen** [comment]
> Why? 1103 has nothing to do with API

To make sure the wallet syncs I was going to run @bokobza's https://gist.github.com/bokobza/8b25a5edeb36f1343372c2d4cc14438c , and need to create wallets for each environment.

Had a chat with @fassadlr and he said it's better to remove `AddMining()` from Stratis.BitcoinD.   And also to rename `MinerController` to `StakingController` as this is what the class is actually doing.

**Comment by ferdeen** [comment]
Cool @fassadlr.  I'll pick this up in the morning.  Cheers.

File: src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj [hunk 2]
```diff
@@ -5,6 +5,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="CSharpFunctionalExtensions" Version="1.10.0" />
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj]: Whats this?

Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.SmartContracts/Stratis.Bitcoin.Features.SmartContracts.csproj]: This was missing from the Stratis.Bitcoin.Features.SmartContracts.csproj file and stopped me from building this project.

Here is the original file from the initial SC commit - https://github.com/stratisproject/StratisBitcoinFullNode/commit/bf7c94d0d9959a6be79910ff072d7e805f9fb01e#diff-e1a79fe58674a2072ea55f05485e7987R7.

**Comment by noescape00** [comment]
>@noescape00 The idea of removing AddMining() didn't work.

The idea was to introduce an attribute that will disable API methods on networks that are not supposed to support them. No need to remove AddMining itself.

**Comment by ferdeen** [comment]
@noescape00 Yes, agreed.  This was a suggestion to try. 

For now can we have this PR merged then I can look at the attribute at a later date.

**Comment by Aprogiena** [comment]
> Why having a suboptimal solution now and fix it later instead of implementing a proper one right away?

In this case I would agree that it is probably worth exploring now. In general it may often be reasonable to go with suboptimal temporary solution.

File: src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+using Stratis.Bitcoin.Controllers;
+using Stratis.Bitcoin.Features.Miner.Interfaces;
+using Stratis.Bitcoin.Features.Miner.Models;
+using Stratis.Bitcoin.Features.Wallet.Interfaces;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.Features.Miner.Controllers
+{
+    /// <summary>
+    /// RPC controller for calls related to PoW mining and PoS minting.
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs]: `/// RPC controller for calls related to PoS minting.`

File: src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    /// </summary>
+    [Controller]
+    public class StakingRpcController : FeatureController
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs]: Shouldn't `route` be specified here?

`[Route("api/[controller]")]`

**Comment by fassadlr** [comment]
PS fix conflicts 👍

File: src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs (hunk 4 above)
Comment (fassadlr) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs]: Oh no it shouldn't, this is the RPC controller...

**Comment by ferdeen** [comment]
What we have done here is tidy up the RPC and API controller around staking/mining.  This allowed us to set the correct controllers for Mining and Staking features.

During this process we realised that all API controllers are discoverable in Swagger.   

So starting StratisD or BitcoinD you will always see all the API controllers (Staking and Mining) if at least one API controller is included within a feature setup.  We will need to tackle this at a later date and focus on sprint tasks.

I'm sorting out the conflicts right now.

File: src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs (hunk 3 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs]: cool

File: src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs (hunk 4 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/StakingRPCController.cs]: The code uses Route for controllers that inherit from `AspNetCore.MVC.Controller` (the API type controllers).

For RPC type controllers these inherit from `FeatureController`.

**Comment by fassadlr** [comment]
@ferdeen please confirm staking can be started via the api calls as per usual.

**Comment by ferdeen** [comment]
**StratisD Staking Test Evidence.**

`GetStakingInfo()`

image

`StartStaking()`

image

`StopStaking()`

image

File: src/Stratis.Bitcoin.Features.Miner/Controllers/StakingApiController.cs [hunk 5]
```diff
@@ -63,9 +59,6 @@ public StakingApiController(IFullNode fullNode, ILoggerFactory loggerFactory, IW
         [HttpGet]
         public IActionResult GetStakingInfo()
         {
-            if (this.fullNode.Network.IsBitcoin())
```
Comment (fassadlr) [inline src/Stratis.Bitcoin.Features.Miner/Controllers/StakingApiController.cs]: Ah dude,  this need to be put back, your test is failing cause you just need to set the network on the mocked FullNode object.

Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Miner/Controllers/StakingApiController.cs]: This was the code that I put in originally to return out of the Staking API methods on the Bitcoin
network.  

I have reverted this change and we have refactored the controllers so the feature setup for BitcoinD and StratisD will use the relevant controllers..

**Comment by TjadenFroyda** [comment]
MvcBuilderExtensions.cs from Features.Api is used to automatically find all of the api endpoints. 

Could the network specific mining/staking endpoints be filtered there?

**Comment by ferdeen** [comment]
@TjadenFroyda - This looks like the right place - or maybe attributes that get checked in MvcBuilderExtensions.cs.  

I'll add your suggestion to #1747.  

Thanks for letting us know.

**Comment by fassadlr** [comment]
@TjadenFroyda perhaps you can look into this for us?

**Comment by TjadenFroyda** [comment]
PR #1842  submitted

**Pull request merged**